internal/engine/interpreter/interpreter.go

```diff
@@ -4336,13 +4338,82 @@ func (ce *callEngine) callNativeFunc(ctx context.Context, m *wasm.ModuleInstance
 			memoryInst.Mux.Unlock()
 			ce.pushValue(uint64(old))
 			frame.pc++
+		case operationKindTailCallReturnCall:
+			f := &functions[op.U1]
+			ce.dropForTailCall(frame, f)
+			if *f == *(frame.f) {
+				// Short-circuit self-recursion.
+				frame.pc = 0
+				continue
+			}
+
+			// The compiler is currently allowing proper tail call only across functions
+			// that belong to the same module; thus, we can overwrite the frame in-place.
+			// For details, see internal/engine/RATIONALE.md
+			frame.f = f
+			frame.base = len(ce.stack)
+			frame.pc = 0
+			body = frame.f.parent.body
+			bodyLen = uint64(len(body))
```

Comment:
also can you share this with non-tailcallindirect and tailcallindirect and share the comment